### PR TITLE
Refactor type_descriptions in the typing env

### PR DESCRIPTION
**Pull request by nchataing**
For new datatype definitions the typer records "declarations", which are mostly direct information obtained from the program source, and "descriptions", which contains lower-level, computed information on data representation (constructor tags, record field offsets).

Before this PR, the type used to define datatype descriptions was

    type type_descriptions =
      constructor_description list * label_description list

which basically assumes that a datatype may have both constructors and labels, which is nonsensical.

The current refactoring uses a variant instead of a product here, following the existing structure of the 'type_kind' field of declarations. This avoids having to use empty lists in impossible configurations, and makes the code less surprising.

### CODE REVIEW

**Comment by gasche** [comment]
@t6s: I wonder if this is something you may be interested in reviewing. It's a sort of a cleanup PR that is not completely different from what you have been working on with @garrigue.

(I supervised @nchataing as he implemented this PR so I'm not independent enough to review it.)

**Comment by t6s** [comment]
I like this PR's improvement on the case analyses on `type_kind`.
The sharing of structures between declarations and descriptions by
introducing a new type (`datatype_kind`) also looks correct.

The increase in the lines of code (by about 40) is apparently a drawback.
However, the cost is mostly paid for pattern matching and value generation
around the new `datatype_kind`, which seems inevitable to achieve the core
value of this PR.

The following type names are now a bit confusing:
* `datatype_kind`:
* `type_kind`
* `type_declaration`
* `type_descriptions`
I would like these to be renamed reflecting the sharing, for example:
* `datatype_kind` -> `type_kind`
* `type_kind` -> `type_declaration(s)`
* `type_declaration` -> ?

Additionally, but almost irrelevantly, to this PR, I noticed that
`find_type_full` is shadowed at toplevel, which seems to let bugs
enter when refactoring the source. Is this an accepted habit in
the source tree of OCaml compiler?

**Comment by gasche** [comment]
Thanks @t6s for your comments!  

> I would like these to be renamed reflecting the sharing, for example:
> - datatype_kind -> type_kind
> - type_kind -> type_declaration(s)
> - type_declaration -> ?

I agree that the naming could be improved, but I am wary of changing too many things at once, in particular types that are used a lot in the codebase and possibly in compiler-libs-using programs. Ideally we would find a naming scheme that would feel nice but, at the same time, not touch the "outer" types like type_declaration.

Here would be a proposal:

- datatype_kind -> type_kind (or actually "datatype_kind" is fine for me, in my mind "datatype" means "variant (possibly extensible) or record or abstract type", the generative/nominal types in OCaml type declarations)
- type_kind -> type_declaration_kind
- type_descriptions -> type_description_kind (or maybe we could just introduce an alias)
- type_declaration stays unchanged 

> Additionally, but almost irrelevantly, to this PR, I noticed that
find_type_full is shadowed at toplevel, which seems to let bugs
enter when refactoring the source. Is this an accepted habit in
the source tree of OCaml compiler?

This is indeed a pattern used in the compiler codebase, but personally I don't like it (especially when the type of the two declarations are the same; otherwise it's harder to confuse the two). I prefer to use a different name for the first, non-public variant, sometimes just `foo_` instead of `foo`. In this case maybe the first `find_type_full` variant could be renamed `find_type_simple`.

**Comment by t6s** [comment]
> I agree that the naming could be improved, but I am wary of changing too many things at once, in particular types that are used a lot in the codebase and possibly in compiler-libs-using programs. Ideally we would find a naming scheme that would feel nice but, at the same time, not touch the "outer" types like type_declaration.

I understand your concern and that policy on renaming sounds reasonable.

> Here would be a proposal:
> 
> * datatype_kind -> type_kind (or actually "datatype_kind" is fine for me, in my mind "datatype" means "variant (possibly extensible) or record or abstract type", the generative/nominal types in OCaml type declarations)
> * type_kind -> type_declaration_kind
> * type_descriptions -> type_description_kind (or maybe we could just introduce an alias)
> * type_declaration stays unchanged

This looks systematic.  It seems good also to me to keep `datatype_kind`, maybe with your explanation added as a comment.

> In this case maybe the first `find_type_full` variant could be renamed `find_type_simple`.

I prefer that.

**Comment by nchataing** [comment]
Thanks for the review @t6s!

I have rebased the PR and made the following changes accordingly :
- the second `find_type_full` was renamed `find_type_data`;
- `datatype_kind` was renamed `type_kind`;
- `type_kind` (in `type_declaration`) was renamed `type_decl_kind`;
- `type_descriptions` (in `env.ml`) was renamed `type_descr_kind`, and an alias for the old name was added.

**Comment by t6s** [comment]
> * the second `find_type_full` was renamed `find_type_data`;
> * `datatype_kind` was renamed `type_kind`;
> * `type_kind` (in `type_declaration`) was renamed `type_decl_kind`;
> * `type_descriptions` (in `env.ml`) was renamed `type_descr_kind`, and an alias for the old name was added.

I am happy with these changes!

(Btw, as I am quite new to the development of ocaml, I am not sure what to do next as a reviewer. Am I supposed to approve this PR?)

**Comment by gasche** [comment]
> Am I supposed to approve this PR?

Yes, that would be nice. Anyone can approve a PR and it gives useful information. The development policy requires that at least a maintainer approves before merging (generally based on their own review, sometimes based on a previous approval), so we also need to ping a maintainer who may be interested in looking at this -- cc @trefis or @Octachron, maybe?

**Review by t6s — APPROVED** [review]

**Review by Octachron — APPROVED** [review]
I agree with the principle of making types narrower in order to fit data.

It was a bit surprising at first that only `Parmatch` clearly benefits from the change. But the `Env` code is only using the bifunctor interface of `type_descriptions`, thus this is not that surprising in retrospect. I wondered if having `bimap: ('l,'r) type_decl -> ('l->'l2) -> ('r->'r2) -> ('l2,'r2) type_decl` (and iter?) would help, but probably not.

**Comment by gasche** [comment]
@Octachron I don't understand what use-cases of `bimap` you have in mind in Env. The code is not so clear to me, but I have the impression that there is a transposition happening (from several types each with labels or constructors, to an environment that lists all labels and list constructors), which does not correspond to a structure-preserving map. Can you elaborate?

(Thanks for the re-review!)

File: typing/env.ml
Comment (trefis) [inline typing/env.ml:1089]: This is inconsistent with:

File: typing/env.ml
Comment (trefis) [inline typing/env.ml]: this.

(I prefer the former)

File: typing/env.ml
Comment (trefis) [inline typing/env.ml]: I don't like this non-indentation style, and I don't think it's used elsewhere in the codebase, certainly not in this file.

File: typing/env.ml
Comment (trefis) [inline typing/env.ml]: Likewise.

**Review by trefis — CHANGES_REQUESTED** [review]
@Octachron having done a serious review and used smart words, I feel like I can just do an *annoying* review, complaining only about indentation.

You're welcome!

File: typing/env.ml
Comment (Octachron) [inline typing/env.ml]: This is `bimap ~record:... ~variant:...`.

File: typing/env.ml
Comment (Octachron) [inline typing/env.ml]: This is another `bimap` call followed by a pattern match to update `env` and a final `bimap` to remove the extra infos. (The current code is probably simpler to follow).

File: typing/env.ml
Comment (gasche) [reply typing/env.ml]: Right. thanks! (I was looking at other parts of `env`.) Indeed one could define `bimap` (I would just call it `map`, given that there is no natural bias for a unimap) and use it here as a refactoring.

File: typing/env.ml
Comment (nchataing) [reply typing/env.ml]: Would
```ocaml
let descrs = match decl.type_kind with
  | Type_variant _ ->
```
be preferable ?
Or rather
```ocaml
let descrs =
  match decl.type_kind with
  | Type_variant _ ->
```

Comment (trefis) [reply typing/env.ml]: I have a personal reference for the second, but either is fine honestly (I believe both are used in the codebase, with the second being perhaps more popular in this particular file)

**Pull request merged**
